File: internal/config/config_test.go
package config

import (
	"os"
	"path/filepath"
	"reflect"
	"testing"
)

func strPtr(s string) *string { return &s }

func defaultLLMBaseURL() *string {
	return strPtr(DefaultLLMBaseURL)
}

func TestLoad(t *testing.T) {
	tests := []struct {
		name      string
		yaml      string
		writeFile bool
		want      Config
		wantErr   bool
	}{
		{
			name:      "full config",
			writeFile: true,
			yaml:      "slack_webhook: https://hooks.slack.com/test\noutput_dir: /tmp/tasks\ngjoll_env: /path/to/sandbox.tf\nsandbox_backend: gjoll\nllm_base_url: \"\"\n",
			want: Config{
				SlackWebhook:   "https://hooks.slack.com/test",
				OutputDir:      "/tmp/tasks",
				SandboxBackend: "gjoll",
				GjollEnv:       "/path/to/sandbox.tf",
				PodmanImage:    "fedora:43",
				AgentBackend:   "opencode",
				LLMBaseURL:     strPtr(""),
				AnthropicKeyFile: "~/.anthropic/api_key",
			},
		},
		{
			name:      "defaults applied",
			writeFile: true,
			yaml:      "slack_webhook: https://hooks.slack.com/test\n",
			want: Config{
				SlackWebhook:   "https://hooks.slack.com/test",
				OutputDir:      "./tasks",
				SandboxBackend: "gjoll",
				GjollEnv:       "./configs/sandbox.tf",
				PodmanImage:    "fedora:43",
				AgentBackend:   "opencode",
				LLMBaseURL:     defaultLLMBaseURL(),
			},
		},
		{
			name:      "empty file uses all defaults",
			writeFile: true,
			yaml:      "",
			want: Config{
				OutputDir:      "./tasks",
				SandboxBackend: "gjoll",
				GjollEnv:       "./configs/sandbox.tf",
				PodmanImage:    "fedora:43",
				AgentBackend:   "opencode",
				LLMBaseURL:     defaultLLMBaseURL(),
			},
		},
		{
			name:      "allowed_repos parsed",
			writeFile: true,
			yaml:      "allowed_repos:\n  - osbuild/osbuild\n  - drellabot/*\n",
			want: Config{
				OutputDir:      "./tasks",
				SandboxBackend: "gjoll",
				GjollEnv:       "./configs/sandbox.tf",
				PodmanImage:    "fedora:43",
				AgentBackend:   "opencode",
				LLMBaseURL:     defaultLLMBaseURL(),
				AllowedRepos:   []string{"osbuild/osbuild", "drellabot/*"},
			},
		},
		{
			name:      "daemon config parsed",
			writeFile: true,
			yaml:      "daemon:\n  poll_interval: \"30s\"\n  allowed_commenters:\n    - alice\n    - bob\n",
			want: Config{
				OutputDir:      "./tasks",
				SandboxBackend: "gjoll",
				GjollEnv:       "./configs/sandbox.tf",
				PodmanImage:    "fedora:43",
				AgentBackend:   "opencode",
				LLMBaseURL:     defaultLLMBaseURL(),
				Daemon: DaemonConfig{
					PollInterval:      "30s",
					AllowedCommenters: []string{"alice", "bob"},
				},
			},
		},
		{
			name:      "profiles config parsed",
			writeFile: true,
			yaml:      "profiles_repo: drellabot/profiles\nprofiles_dir: /tmp/profiles\n",
			want: Config{
				OutputDir:      "./tasks",
				SandboxBackend: "gjoll",
				GjollEnv:       "./configs/sandbox.tf",
				PodmanImage:    "fedora:43",
				AgentBackend:   "opencode",
				LLMBaseURL:     defaultLLMBaseURL(),
				ProfilesRepo:   "drellabot/profiles",
				ProfilesDir:    "/tmp/profiles",
			},
		},
		{
			name:      "agent config parsed",
			writeFile: true,
			yaml:      "agent:\n  max-budget-usd: 100\n  warn-budget-usd: 30\n  critical-budget-usd: 50\n",
			want: Config{
				OutputDir:      "./tasks",
				SandboxBackend: "gjoll",
				GjollEnv:       "./configs/sandbox.tf",
				PodmanImage:    "fedora:43",
				AgentBackend:   "opencode",
				LLMBaseURL:     defaultLLMBaseURL(),
				Agent: AgentConfig{
					MaxBudgetUSD:      100,
					WarnBudgetUSD:     30,
					CriticalBudgetUSD: 50,
				},
			},
		},
		{
			name:      "agent config partial",
			writeFile: true,
			yaml:      "agent:\n  max-budget-usd: 50\n",
			want: Config{
				OutputDir:      "./tasks",
				SandboxBackend: "gjoll",
				GjollEnv:       "./configs/sandbox.tf",
				PodmanImage:    "fedora:43",
				AgentBackend:   "opencode",
				LLMBaseURL:     defaultLLMBaseURL(),
				Agent: AgentConfig{
					MaxBudgetUSD: 50,
				},
			},
		},
		{
			name:      "invalid yaml",
			writeFile: true,
			yaml:      "{{invalid",
			wantErr:   true,
		},
		{
			name:      "missing file",
			writeFile: false,
			wantErr:   true,
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			dir := t.TempDir()
			path := filepath.Join(dir, "config.yaml")
			if tt.writeFile {
				if err := os.WriteFile(path, []byte(tt.yaml), 0644); err != nil {
					t.Fatal(err)
				}
			}

			got, err := Load(path)
			if tt.wantErr {
				if err == nil {
					t.Fatal("expected error, got nil")
				}
				return
			}
			if err != nil {
				t.Fatalf("unexpected error: %v", err)
			}
			if !reflect.DeepEqual(*got, tt.want) {
				t.Errorf("got %+v, want %+v", *got, tt.want)
			}
		})
	}
}

func TestUsesLocalLLM(t *testing.T) {
	cfg := &Config{LLMBaseURL: defaultLLMBaseURL()}
	if !cfg.UsesLocalLLM() {
		t.Fatal("expected local LLM with default URL")
	}

	empty := ""
	cfg.LLMBaseURL = &empty
	if cfg.UsesLocalLLM() {
		t.Fatal("expected cloud mode with empty llm_base_url")
	}
}

func TestLocalLLMHostPort(t *testing.T) {
	url := "http://127.0.0.1:11434/v1"
	cfg := &Config{LLMBaseURL: &url}

	port, err := cfg.LocalLLMHostPort()
	if err != nil {
		t.Fatal(err)
	}
	if port != 11434 {
		t.Fatalf("port = %d, want 11434", port)
	}

	got, err := cfg.GjollLLMBaseURL()
	if err != nil {
		t.Fatal(err)
	}
	if got != "http://127.0.0.1:11434/v1" {
		t.Fatalf("GjollLLMBaseURL() = %q", got)
	}
}

func TestAgentOptionsGjollLocalLLM(t *testing.T) {
	url := "http://127.0.0.1:11434/v1"
	cfg := &Config{
		SandboxBackend: "gjoll",
		LLMBaseURL:     &url,
		LLMModel:       "test-model",
	}
	opts := cfg.AgentOptions()
	if opts.LLMBaseURL != "http://127.0.0.1:11434/v1" {
		t.Fatalf("LLMBaseURL = %q, want gjoll proxy URL", opts.LLMBaseURL)
	}
	if opts.LLMModel != "test-model" {
		t.Fatalf("LLMModel = %q", opts.LLMModel)
	}
}

func TestRepoAllowed(t *testing.T) {
	tests := []struct {
		name         string
		allowedRepos []string
		repo         string
		want         bool
	}{
		{
			name:         "exact match",
			allowedRepos: []string{"osbuild/osbuild"},
			repo:         "osbuild/osbuild",
			want:         true,
		},
		{
			name:         "wildcard match",
			allowedRepos: []string{"drellahq/*"},
			repo:         "drellahq/orchestrator",
			want:         true,
		},
		{
			name:         "no match",
			allowedRepos: []string{"osbuild/osbuild"},
			repo:         "evil/repo",
			want:         false,
		},
		{
			name:         "empty list denies all",
			allowedRepos: nil,
			repo:         "osbuild/osbuild",
			want:         false,
		},
		{
			name:         "wildcard does not cross slash",
			allowedRepos: []string{"org/*"},
			repo:         "org/a/b",
			want:         false,
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			cfg := &Config{AllowedRepos: tt.allowedRepos}
			if got := cfg.RepoAllowed(tt.repo); got != tt.want {
				t.Errorf("RepoAllowed(%q) = %v, want %v", tt.repo, got, tt.want)
			}
		})
	}
}

File: src/config.test.ts
import { describe, it, expect } from "vitest";
import * as fs from "node:fs";
import * as path from "node:path";
import * as os from "node:os";
import { load, repoAllowed, type Config } from "./config.js";

function withConfig(content: string): string {
  const dir = fs.mkdtempSync(path.join(os.tmpdir(), "config-test-"));
  const p = path.join(dir, "config.yaml");
  fs.writeFileSync(p, content);
  return p;
}

describe("load", () => {
  it("full config", () => {
    const p = withConfig(
      "slack_webhook: https://hooks.slack.com/test\noutput_dir: /tmp/tasks\ngjoll_env: /path/to/sandbox.tf\n"
    );
    const cfg = load(p);
    expect(cfg.slack_webhook).toBe("https://hooks.slack.com/test");
    expect(cfg.output_dir).toBe("/tmp/tasks");
    expect(cfg.gjoll_env).toBe("/path/to/sandbox.tf");
  });

  it("defaults applied", () => {
    const p = withConfig("slack_webhook: https://hooks.slack.com/test\n");
    const cfg = load(p);
    expect(cfg.output_dir).toBe("./tasks");
    expect(cfg.gjoll_env).toBe("./configs/sandbox.tf");
  });

  it("empty file uses all defaults", () => {
    const p = withConfig("");
    const cfg = load(p);
    expect(cfg.output_dir).toBe("./tasks");
    expect(cfg.gjoll_env).toBe("./configs/sandbox.tf");
    expect(cfg.slack_webhook).toBeUndefined();
  });

  it("allowed_repos parsed", () => {
    const p = withConfig(
      "allowed_repos:\n  - osbuild/osbuild\n  - drellabot/*\n"
    );
    const cfg = load(p);
    expect(cfg.allowed_repos).toEqual(["osbuild/osbuild", "drellabot/*"]);
  });

  it("daemon config parsed", () => {
    const p = withConfig(
      'daemon:\n  poll_interval: "30s"\n  allowed_commenters:\n    - alice\n    - bob\n'
    );
    const cfg = load(p);
    expect(cfg.daemon.poll_interval).toBe("30s");
    expect(cfg.daemon.allowed_commenters).toEqual(["alice", "bob"]);
  });

  it("invalid yaml throws", () => {
    const p = withConfig("{{invalid");
    expect(() => load(p)).toThrow();
  });

  it("missing file throws", () => {
    expect(() => load("/nonexistent/config.yaml")).toThrow();
  });
});

describe("repoAllowed", () => {
  it("exact match", () => {
    const cfg = { allowed_repos: ["osbuild/osbuild"] } as Config;
    expect(repoAllowed(cfg, "osbuild/osbuild")).toBe(true);
  });

  it("wildcard match", () => {
    const cfg = { allowed_repos: ["drellabot/*"] } as Config;
    expect(repoAllowed(cfg, "drellabot/orchestrator")).toBe(true);
  });

  it("no match", () => {
    const cfg = { allowed_repos: ["osbuild/osbuild"] } as Config;
    expect(repoAllowed(cfg, "evil/repo")).toBe(false);
  });

  it("empty list denies all", () => {
    const cfg = { allowed_repos: [] } as Config;
    expect(repoAllowed(cfg, "osbuild/osbuild")).toBe(false);
  });
});
